docs/adr/009_discovery.md

```diff
@@ -0,0 +1,81 @@
+| Bidirectional connection | Bidirectional connection (e.g. WebSocket) is permanently open between the discovery service endpoint and each LNS. |                                           | - Sensitive to transient connection drops<br />- Not necessarily a good indicator of service health (LNS can be alive but not reach discovery service) |
+| D2C message              | LNS sends a periodic D2C message to indicate that it's still alive | - Also tests D2C delivery capabilities    | - Tests D2C delivery capabilities (potential problems in offline scenario) |
+| Module twin status       | Discovery service relies on IoT Hub module twin for detecting module outages | - Reuses the IoT Edge monitoring strategy | From docs: he reported properties of a runtime module could be stale if an IoT Edge device gets disconnected from its IoT hub. |
+| Ping discovery service   | Each LNS pings the discovery service on a regular basis to indicate that it's still alive. |                                           | - Time between crash of LNS and detection of crash           |
+| State update             | LNS updates state on a regular basis, which can be queried by the discovery service, e.g. using Azure Storage blob lease management (and lease renewal) |                                           | - Time between crash of LNS and detection of crash<br />- Complexity |
+| Health endpoint on LNS   | Each LNS has a health-check endpoint, that can be queried by the discovery service |                                           | - Does not necessarily work for on-premises deployments of LNS |
+| No health checks at all  | The discovery service relies on the implicit knowledge that if an LBS issues another request to the discovery service, the connection to the LNS failed | - Simple                                  | - Depends on how long the station waits before back-off (and how many LNS are down)<br />- Takes around two minutes (20 seconds per default on LBS 2.0.6) on a concentrator to reconnect to the discovery endpoint if the LNS is unavailable. |
+
+We should keep in mind that we can also use a combination of the strategies and use an incremental approach again: start with a simple health detection strategy, and allow configuring a different health detection strategy.
+
+The selection of a health check strategy should take into account how many upstream messages can potentially be lost, depending on how long the downtime is. To investigate this further, we ran a spike where we simulated the LNS going down and then recovering after different delays, starting at 10 sec. In our experiment, the leaf device was configured to send unconfirmed messages every 5 sec. We found that in case the downtime was 1 min or shorter, we were able to recover all messages that were broadcasted in the meantime. If the downtime was over 1 min, all messages were lost.
```

Comment:
I would remove this paragraph as it might be misleading. As soon as we are going to connect to /router-info, we lose all the messages.

Reply:
I think it's important to mention that the longer we take to reconnect an LBS to an LNS, the more messages the devices send and the more messages we can potentially lose (even though this may be obvious).

I rephrased it with 15a7bb958ce61dc6a7cb58e15d68a9d245a6d409 to make it clearer that the time should be taken into account not because of the behavior of the LBS, but simply because the more time passes the more messages the end devices send. Please resolve if it's clearer in the latest version.